dlrover/python/master/monitor/metric_context.py

```diff
@@ -0,0 +1,317 @@
+# Copyright 2024 The DLRover Authors. All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+from typing import Dict
+
+from dlrover.python.common.constants import GpuMetricType, NpuMetricType
+from dlrover.python.common.global_context import Context
+from dlrover.python.common.singleton import Singleton
+
+_dlrover_context = Context.singleton_instance()
+
+
+class XpuMetric(metaclass=ABCMeta):
+    """XPU Metric
+    Attributes:
+        type(string): xpu type
+    """
+
+    def __init__(self, xpu_type):
+        self.type = xpu_type
+
+    @abstractmethod
+    def set_metric(self, key, value):
+        pass
+
+    @abstractmethod
+    def get_metric(self, key):
+        pass
+
+
+class GpuMetric(XpuMetric):
+    """GPU Metric
+    Attributes:
+        gpu_free_mem(int, MB): free gpu memory
```

Comment:
These kind of metric(used/free/total/) should be unified with npu type.

Reply:
npu and gpu metrics are different from raw data